README.md
# java-api-wrapper

OAuth2 SoundCloud API wrapper written in Java ([javadoc][]).

It is simple to use and requires a minimum of external dependencies (compared to
the OAuth1 wrapper) so should be easily embeddable in both desktop and
mobile applications.

## Android

The wrapper works well on Android (although it has no dependencies on it)
since it is an extraction from our [SoundCloud Android][] codebase. However, if
all you want is to share sounds from your own application we recommend to check out the
[Android Sharing Kit][] which delegates all the hard work to the SoundCloud
app and makes use of Android's [intent][] framework.

## Usage

Create a wrapper instance:

```java
ApiWrapper wrapper = new ApiWrapper("client_id", "client_secret", null, null);
```

Obtain a token:

```java
wrapper.login("username", "password");
```

Execute a request:

```java
HttpResponse resp = wrapper.get(Request.to("/me"));
```

Update a resource:

```java
HttpResponse resp =
      wrapper.put(Request.to("/me")
             .with("user[full_name]", "Che Flute",
                   "user[website]",   "http://cheflute.com")
             .withFile("user[avatar_data]", new File("flute.jpg")));
```

## Refresh tokens

OAuth2 access tokens are only valid for a certain amount of time (usually 1h)
and need to be refreshed when they become stale. The wrapper automatically
refreshes the token and retries the request so an API client usually does not
need to care about this fact. If the client is interested (possibly to persist
the updated token) it can register a listener with the wrapper.

## Non-expiring access tokens (only applies to version 1.0.1+)

Expiring access tokens provide more security but also add more complexity to
the authentication process. If you don't want to use them you can request
non-expiring tokens by specifying the scope "non-expiring" when exchanging the
tokens:

```java
Token token = wrapper.login("username", "password", Token.SCOPE_NON_EXPIRING);
```

The resulting token will be valid until revoked manually.

For the `authorization_code` grant type you need to request the scope like so:

```java
URI uri = wrapper.authorizationCodeUrl(Endpoints.CONNECT, Token.SCOPE_NON_EXPIRING);
// open uri in browser / WebView etc.
```

## Login via Facebook

Please see [FacebookConnect][] for an example of this login flow.

## Requirements

The wrapper depends on [Apache HttpClient][] (including the [HttpMime][]
module) and [json-java][]. The Android SDK already comes with these two
libraries so you don't need to include them when using the wrapper there.

## Build + Test

The project uses the groovy-based build system [gradle][] (version 1.x):

    $ brew update && brew install gradle (OSX+homebrew, check website for other OS)
    $ git clone git://github.com/soundcloud/java-api-wrapper.git
    $ cd java-api-wrapper
    $ gradle jar  # build jar file (build/libs/java-api-wrapper-1.x.x.jar)
    $ gradle test # run tests

You don't have to use gradle - the repo also contains a `pom.xml` file which
can be used to build and test the project with [Apache Maven][] (`mvn install`).

Jar files are available in the Github [download section][downloads] and on
sonatype.org / maven ([snapshots][], [releases][], [maven-central][]).

## Examples

The wrapper ships with a few examples in `src/examples/java`:

  * [CreateWrapper][] creates a wrapper and obtains an access token using
  login / password.
  * [GetResource][] performs a GET request for a resource and prints the
  JSON result.
  * [PostResource][] performs a POST request to create a resource and prints the
  JSON result
  * [PutResource][] performs a PUT request to update a resource and prints the
  JSON result
  * [UploadFile][] uploads a file to SoundCloud.
  * [FacebookConnect][] obtain an access token via Facebook login

You can use gradle tasks to compile and run these examples with one command.

First create a wrapper and remember to substitute all credentials with real ones
([register an app][register-app] if you need client_id/secret):

    # with gradle
    $ gradle createWrapper -Pclient_id=my_client_id \
        -Pclient_secret=mys3cr3t \
        -Plogin=api-testing \
        -Ppassword=testing

    # with plain java
    $ java -classpath java-api-wrapper-1.x.y-all.jar \
        com.soundcloud.api.examples.CreateWrapper \
        my_client_id mys3cr3t api-testing testing

Output:

    got token from server: Token{
      access='0000000KNYbSTHKNZC2tq7Epkgxvgmhu',
      refresh='0000000jd4YCL0vCuKf6UtPsS6Ahd0wc', scope='null',
      expires=Mon May 02 17:35:15 CEST 2011}

    wrapper serialised to wrapper.ser

With the wrapper and all tokens serialised to `wrapper.ser` you can run the
other examples.

GET a resource:

    $ gradle getResource -Presource=/me
    (java -classpath java-api-wrapper-1.x.y-all.jar \
        com.soundcloud.api.examples.GetResource /me)

Output:

    GET /me
    {
        "username": "testing",
        "city": "Berlin"
        ...
    }

PUT a resource:

    $ gradle putResource -Presource=/me -Pcontent='{ "user": { "city": "Testor" } }' -PcontentType=application/json

Output:

    PUT /me
    {
        "username": "testing",
        "city": "Testor"
        ...
    }

POST a resource:

    $ gradle postResource -Presource=/playlists -Pcontent='{ "playlist": { "title": "Test" } }' -PcontentType=application/json

Output:

    POST /playlists
    {
        "title": "Test",
        "artwork_url": null,
        ...
    }

Upload a file:

    $ gradle uploadFile \
            -Pfile=src/test/resources/com/soundcloud/api/hello.aiff
      (java -classpath java-api-wrapper-1.x.y-all.jar \
        com.soundcloud.api.examples.UploadFile ...)

Output:

    Uploading src/test/resources/com/soundcloud/api/hello.aiff
    .............................................
    201 Created https://api.sandbox-soundcloud.com/tracks/2100052
    {
        "artwork_url": null,
        ...
    }

You can add the debug flag (`-d`) to gradle to get some extra HTTP logging:

    $ gradle getResource -Presource=/me -d

    [DEBUG] DefaultClientConnection - Sending request: GET /me HTTP/1.1
    [DEBUG] headers - >> GET /me HTTP/1.1
    [DEBUG] headers - >> Authorization: OAuth 0000000ni3Br147FO7Cj5XotAy
    ...

Note that while the example code uses standard Java serialization to persist
state across calls you should probably use a different mechanism in your app.

## Note on Patches/Pull Requests

  * Fork the project.
  * Make your feature addition or bug fix.
  * Add tests for it.
  * Commit, do not mess with buildfile, version, or history.
  * Send a pull request. Bonus points for topic branches.

If you want to work on the code in an IDE instead of a text editor you can
easily create project files with gradle:

    $ gradle idea     # Intellij IDEA
    $ gradle eclipse  # Eclipse

## Credits / License

The API is based on [urbanstew][]'s [soundcloudapi-java][] project.

Includes portions of code (c) 2010 Xtreme Labs and Pivotal Labs and (c) 2009 urbanSTEW.

See LICENSE for details.

[![Build Status](https://secure.travis-ci.org/soundcloud/java-api-wrapper.png?branch=master)](http://travis-ci.org/soundcloud/java-api-wrapper)

[gradle]: http://www.gradle.org/
[urbanstew]: http://urbanstew.org/
[Apache HttpClient]: http://hc.apache.org/httpcomponents-client-ga/
[HttpMime]: http://hc.apache.org/httpcomponents-client-ga/httpmime
[json-java]: http://json.org/java/
[javadoc]: http://soundcloud.github.com/java-api-wrapper/javadoc/1.3.0/com/soundcloud/api/package-summary.html
[soundcloudapi-java]: http://code.google.com/p/soundcloudapi-java/
[soundcloudapi-java-annouce]: http://blog.soundcloud.com/2010/01/08/java-wrapper/
[CreateWrapper]: https://github.com/soundcloud/java-api-wrapper/blob/master/src/examples/java/com/soundcloud/api/examples/CreateWrapper.java
[GetResource]: https://github.com/soundcloud/java-api-wrapper/blob/master/src/examples/java/com/soundcloud/api/examples/GetResource.java
[PutResource]: https://github.com/soundcloud/java-api-wrapper/blob/master/src/examples/java/com/soundcloud/api/examples/PutResource.java
[PostResource]: https://github.com/soundcloud/java-api-wrapper/blob/master/src/examples/java/com/soundcloud/api/examples/PostResource.java
[UploadFile]: https://github.com/soundcloud/java-api-wrapper/blob/master/src/examples/java/com/soundcloud/api/examples/UploadFile.java
[FacebookConnect]: https://github.com/soundcloud/java-api-wrapper/blob/master/src/examples/java/com/soundcloud/api/examples/FacebookConnect.java
[SoundCloud Android]: https://play.google.com/store/apps/details?id=com.soundcloud.android
[register-app]: http://soundcloud.com/you/apps/new
[Apache Maven]: http://maven.apache.org/
[downloads]: https://github.com/soundcloud/java-api-wrapper/archives/master
[snapshots]: https://oss.sonatype.org/content/repositories/snapshots/com/soundcloud/java-api-wrapper/
[releases]: https://oss.sonatype.org/content/repositories/releases/com/soundcloud/java-api-wrapper/
[maven-central]: http://repo1.maven.org/maven2/com/soundcloud/java-api-wrapper/
[Android Sharing Kit]: https://github.com/soundcloud/android-intent-sharing/wiki
[android-token-sharing]: https://github.com/soundcloud/android-token-sharing
[intent]: http://developer.android.com/reference/android/content/Intent.html


CHANGES.md
## 1.3.1 2013-10-02

  * Always include client_id in requests [23]
  * Added support for setting multiple values per query parameter [5e830db]
  * Do not default to using env.sslResourceHost for resolving streams [21]

## 1.3.0 2013-04-22

  * Default to UTF-8 for string payloads
  * Include client_id for unauthenticated requests
  * Support gzip encoding
  * Fixed non-english request formatting issue
  * Preserve query parameters for 'withContent`

## 1.2.1 2013-01-10

  * Added ApiResponseException to wrap HTTP error codes when logging in
  * Fixed: Token.java: JSONObject["expires_in"] not a string [8]
  * Fixed: NoSuchMethodError on CharsetUtil.getCharset() [7]
  * Fixed: impossible to access a public resource without being logged [6]

## 1.2.0 2012-11-05

  * Handle more broken HTTP client behaviour
  * Remove sandbox environment (it is no longer supported)

## 1.1.2 2012-10-04

  * Compatibility fixes with broken httpclient versions used in Android pre-gingerbread
  * Add support for default parameters

## 1.1.1 2012-04-30

  * Respect system proxy settings
  * Added skip_logging parameter to resolveStreamUrl
  * Added workarounds for some HTTPClient bugs
  * Fixed some Android compatibility problems (IOException constructor)

## 1.1.0 2011-11-09

  * Support httpmime-4.1.x (GH-2)
  * Support for conditional GETs
  * Support for Range requests
  * added CloudApi#resolveStreamUrl(String)
  * added CloudApi#getHttpClient()
  * Changed the handling of max connections per route
  * Added some endpoints
  * Added PostResource example
  * Added support for HEAD requests
  * Added stream resolving
  * Added Facebook login example

## 1.0.1 2011-07-04

  * Support for non-expiring scope
  * Cancellation of uploads
  * Added Request#withFile(String, byte[]), Request#withFile(String, ByteBuffer)
  * Added Request#withEntity(HttpEntity) and Request#withContent(String, String)
  * Added PutResource example
  * Added setting to change the default content type for requests
    (CloudAPI#setDefaultContentType(String))

## 1.0.0 2011-05-19

  * Initial release


[5e830db]: https://github.com/soundcloud/java-api-wrapper/commit/5e830db9ebcb03869b3ad146197738f100e84e09
[23]: https://github.com/soundcloud/java-api-wrapper/issues/23
[21]: https://github.com/soundcloud/java-api-wrapper/pull/21
[6]: https://github.com/soundcloud/java-api-wrapper/issues/6
[7]: https://github.com/soundcloud/java-api-wrapper/issues/7
[8]: https://github.com/soundcloud/java-api-wrapper/issues/8


src/main/java/com/soundcloud/api/CloudAPI.java
package com.soundcloud.api;

import org.apache.http.HttpHost;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpUriRequest;

import java.io.IOException;
import java.net.URI;

/**
 * Interface with SoundCloud, using OAuth2.
 *
 * This is the interface, for the implementation see ApiWrapper.
 * @see ApiWrapper
 */
public interface CloudAPI {
    // OAuth2 parameters
    String GRANT_TYPE       = "grant_type";
    String CLIENT_ID        = "client_id";
    String CLIENT_SECRET    = "client_secret";
    String USERNAME         = "username";
    String REDIRECT_URI     = "redirect_uri";
    String CODE             = "code";
    String RESPONSE_TYPE    = "response_type";
    String SCOPE            = "scope";

    // standard oauth2 grant types
    String PASSWORD           = "password";
    String AUTHORIZATION_CODE = "authorization_code";
    String REFRESH_TOKEN      = "refresh_token";
    String CLIENT_CREDENTIALS = "client_credentials";

    // custom
    String OAUTH1_TOKEN_GRANT_TYPE  = "oauth1_token"; // soundcloud
    String FACEBOOK_GRANT_TYPE      = "urn:soundcloud:oauth2:grant-type:facebook&access_token="; // oauth2 extension
    String GOOGLE_PLUS_GRANT_TYPE   = "urn:soundcloud:oauth2:grant-type:google_plus&access_token=";

    // other constants
    String REALM              = "SoundCloud";
    String OAUTH_SCHEME       = "oauth";
    String VERSION            = "1.3.0";
    String USER_AGENT         = "SoundCloud Java Wrapper ("+VERSION+")";

    /**
     * Request a token using <a href="http://tools.ietf.org/html/draft-ietf-oauth-v2-10#section-4.1.2">
     * Resource Owner Password Credentials</a>.
     *
     * @param username SoundCloud username
     * @param password SoundCloud password
     * @param scopes   the desired scope(s), or empty for default scope
     * @return a valid token
     * @throws com.soundcloud.api.CloudAPI.InvalidTokenException
     *                     invalid token
     * @throws IOException In case of network/server errors
     */
    Token login(String username, String password, String... scopes) throws IOException;


    /**
     * Request a token using <a href="http://tools.ietf.org/html/draft-ietf-oauth-v2-10#section-4.1.1">
     * Authorization Code</a>, requesting a default scope.
     *
     * @param code the authorization code
     * @param scopes   the desired scope(s), or empty for default scope
     * @return a valid token
     * @throws com.soundcloud.api.CloudAPI.InvalidTokenException invalid token
     * @throws IOException In case of network/server errors
     */
    Token authorizationCode(String code, String... scopes) throws IOException;


    /**
     * Request a "signup" token using <a href="http://tools.ietf.org/html/draft-ietf-oauth-v2-15#section-4.4">
     * Client Credentials</a>.
     *
     * Note that this token is <b>not</b> set as the current token in the wrapper - it should only be used
     * for one request (typically the signup / user creation request).
     * Also note that not all apps are allowed to request this token type (the wrapper throws
     * InvalidTokenException in this case).
     *
     * @param scopes   the desired scope(s), or empty for default scope
     * @return a valid token
     * @throws IOException IO/Error
     * @throws com.soundcloud.api.CloudAPI.InvalidTokenException if requested scope is not available
     */
    Token clientCredentials(String... scopes) throws IOException;


    /**
     * Request a token using an <a href="http://tools.ietf.org/html/draft-ietf-oauth-v2-22#section-4.5">
     * extension grant type</a>.
     * @param grantType
     * @param scopes
     * @return
     * @throws IOException
     */
    Token extensionGrantType(String grantType, String... scopes) throws IOException;

    /**
     * Tries to refresh the currently used access token with the refresh token.
     * If successful the API wrapper will have the new token set already.
     * @return a valid token
     * @throws IOException in case of network problems
     * @throws com.soundcloud.api.CloudAPI.InvalidTokenException invalid token
     * @throws IllegalStateException if no refresh token present
     */
    Token refreshToken() throws IOException;

    /**
     * This method should be called when the token was found to be invalid.
     * Also replaces the current token, if there is one available.
     * @return an alternative token, or null if none available
     *         (which indicates that a refresh could be tried)
     */
    Token invalidateToken();

    /**
       * @param request resource to HEAD
       * @return the HTTP response
       * @throws IOException IO/Error
       */
    HttpResponse head(Request request) throws IOException;

    /**
     * @param request resource to GET
     * @return the HTTP response
     * @throws IOException IO/Error
     */
    HttpResponse get(Request request) throws IOException;

    /**
     * @param request resource to POST
     * @return the HTTP response
     * @throws IOException IO/Error
     */
    HttpResponse post(Request request) throws IOException;

    /**
     * @param request resource to PUT
     * @return the HTTP response
     * @throws IOException IO/Error
     */
    HttpResponse put(Request request) throws IOException;

    /**
     * @param request resource to DELETE
     * @return the HTTP response
     * @throws IOException IO/Error
     */
    HttpResponse delete(Request request) throws IOException;

    /**
     * @return the used httpclient
     */
    HttpClient getHttpClient();

    /**
     * Generic execute method, with added workarounds for various HTTPClient bugs.
     *
     * @param target the target host (can be null)
     * @param request the request
     * @return the HTTP response
     * @throws IOException network errors
     * @throws BrokenHttpClientException in case of HTTPClient framework bugs
     */
    HttpResponse safeExecute(HttpHost target, HttpUriRequest request) throws IOException;

    /**
     * Resolve the given SoundCloud URI
     *
     * @param uri SoundCloud model URI, e.g. http://soundcloud.com/bob
     * @return the id
     * @throws IOException network errors
     * @throws ResolverException if object could not be resolved
     */
    long resolve(String uri) throws IOException;

    /**
     * Resolve the given SoundCloud stream URI
     *
     * @param uri SoundCloud stream URI, e.g. https://api.soundcloud.com/tracks/25272620/stream
     * @param skipLogging skip logging the play of this track (client needs
     *        {@link com.soundcloud.api.Token#SCOPE_PLAYCOUNT})
     * @return the resolved stream
     * @throws IOException network errors
     * @throws com.soundcloud.api.CloudAPI.ResolverException resolver error (invalid status etc)
     */
    Stream resolveStreamUrl(String uri, boolean skipLogging) throws IOException;

    /** @return the current token */
    Token getToken();

    /** @param token the token to be used */
    void setToken(Token token);

    /**
     * Registers a listener. The listener will be informed when an access token was found
     * to be invalid, and when the token had to be refreshed.
     * @param listener token listener
     */
    void setTokenListener(TokenListener listener);

    /**
     * Request login via authorization code
     * After login, control will go to the redirect URI (wrapper specific), with
     * one of the following query parameters appended:
     * <ul>
     * <li><code>code</code> in case of success, this will contain the code used for the
     *     <code>authorizationCode</code> call to obtain the access token.
     * <li><code>error</code> in case of failure, this contains an error code (most likely
     * <code>access_denied</code>).
     * </ul>
     * @param  options auth endpoint to use (leave out for default), requested scope (leave out for default)
     * @return the URI to open in a browser/WebView etc.
     * @see CloudAPI#authorizationCode(String, String...)
     */
    URI authorizationCodeUrl(String... options);

    /**
     * Changes the default content type sent in the "Accept" header.
     * If you don't set this it defaults to "application/json".
     *
     * @param contentType the request mime type.
     */
    void setDefaultContentType(String contentType);
    void setDefaultAcceptEncoding(String encoding);

    /**
     * Interested in changes to the current token.
     */
    interface TokenListener {
        /**
         * Called when token was found to be invalid
         * @param token the invalid token
         * @return a cached token if available, or null
         */
        Token onTokenInvalid(Token token);

        /**
         * Called when the token got successfully refreshed
         * @param token      the refreshed token
         */
        void onTokenRefreshed(Token token);
    }

    /**
     * Thrown when token is not valid.
     */
    class InvalidTokenException extends IOException {
        private static final long serialVersionUID = 1954919760451539868L;

        /**
         * @param code the HTTP error code
         * @param status the HTTP status, or other error message
         */
        public InvalidTokenException(int code, String status) {
            super("HTTP error:" + code + " (" + status + ")");
        }
    }

    /**
     * Thrown if resolving the audio stream of a SoundCloud sound fails.
     */
    class ResolverException extends ApiResponseException {

        public ResolverException(String s, HttpResponse resp) {
            super(resp, s);
        }

        public ResolverException(Throwable throwable, HttpResponse response) {
            super(throwable, response);
        }
    }

    /**
     * Thrown if the service API responds in error. The HTTP status code can be obtained via {@link #getStatusCode()}.
     */
    class ApiResponseException extends IOException {
        private static final long serialVersionUID = -2990651725862868387L;

        public final HttpResponse response;

        public ApiResponseException(HttpResponse resp, String error) {
            super(resp.getStatusLine().getStatusCode() + ": [" + resp.getStatusLine().getReasonPhrase() + "] "
                    + (error != null ? error : ""));
            this.response = resp;
        }

        public ApiResponseException(Throwable throwable, HttpResponse response) {
            super(throwable == null ? null : throwable.toString());
            initCause(throwable);
            this.response = response;
        }

        public int getStatusCode() {
            return response.getStatusLine().getStatusCode();
        }

        @Override
        public String getMessage() {
            return super.getMessage()+" "+(response != null ? response.getStatusLine() : "");
        }
    }

    class BrokenHttpClientException extends IOException {
        private static final long serialVersionUID = -4764332412926419313L;

        BrokenHttpClientException(Throwable throwable) {
            super(throwable == null ? null : throwable.toString());
            initCause(throwable);
        }
    }
}
